resilience4j-kotlin/src/main/kotlin/io/github/resilience4j/kotlin/circuitbreaker/CircuitBreaker.kt

```diff
@@ -36,12 +33,8 @@ suspend fun <T> CircuitBreaker.executeSuspendFunction(block: suspend () -> T): T
         onResult(durationInNanos, TimeUnit.NANOSECONDS, result)
         return result
     } catch (exception: Throwable) {
-        if (isCancellation(coroutineContext, exception)) {
```

Comment:
I think we cannot just remove this code snippet.
There were could reasons to release a permissions when a coroutine was cancelled.

I assume sometimes you want to handle a `CancellationException` and release a permission and 
sometimes you want to handle a `HttpRequestTimeoutException` and track it as an error.

Reply:
I think this undocumented feature breaks the semantics/API of the circuit breaker. I was caught by
surprise when exceptions were silently swallowed. Shouldn't you use the configuration option `ignoreExceptions` if you wanted this behaviour?

Reply:
No, this release permission functionality is needed when a call is done in HALF_OPEN state.
Every call in HALF_OPEN state decreases the number of permitted calls (`permittedNumberOfCalls
InHalfOpenState`). But if someone cancels the call, we want to increase the number of permitted calls again.

It's implemented similar in Futures, RxJava and Reactor:

Future:
https://github.com/resilience4j/resilience4j/blob/master/resilience4j-circuitbreaker/src/main/java/io/github/resilience4j/circuitbreaker/CircuitBreaker.java#L1088-L1091

Reactor:
https://github.com/resilience4j/resilience4j/blob/master/resilience4j-reactor/src/main/java/io/github/resilience4j/reactor/circuitbreaker/operator/CircuitBreakerSubscriber.java#L73-L81

But in Java a `TimeoutException` does not inherit from a `CancellationException`.

Reply:
But I agree the exception should not be swalloed but rethrown,

Reply:
Not sure about it. It feels wrong that you need to add `CancellationException` to the list of ignored exceptions so that the number of permitted calls is decreased when the CircuitBreaker is in HALF_OPEN state. The idea was that request cancellation works without additional configuration.
It's not neccessary for Futures, RxJava and Reactor. I don't want to introduce this just for Kotlin Ktor.  It would also be a non-backwards compatible change for resilience4j-kotlin. 

It also feels strange that `HttpRequestTimeoutException` inherits from `CancellationException`. What's the reasoning behind it? Does the exception cancel the coroutine which is executing the call using the Ktor client library?

Reply:
But you cant make it work without additional configuration since it is totally dependent on the context if a cancellation is an error, success or should be ignored? I dont think you can always assume that this should be treated as ignore as it is implemented now.  And `HttpRequestTimeoutException` from Ktor Client is a good example of that. A http request is cancelled if it takes too long time, and in my context I want to count that as an error so the CircuitBreaker trips if too many requests times out. The exception is given as a detailed cancellation cause, see https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/-job/cancel.html

Reply:
For backwards-compatibility we could add a new extension function instead of modifying the existing one? Wdyt?

Reply:
Yes, that is OK for me, but then we are left with one of the hard problems in computer science: naming 😄

Reply:
`executeSuspendFunctionAndIgnoreCancellation` ?

Reply:
I think that is a bit confusing since it actually describes the current implementation of `executeSuspendFunction`? Maybe `executeSuspendFunctionAndRecordCancellationError`could work?

Reply:
Or `executeSuspendFunctionAndRecordAllThrowables`?

Reply:
I prefer `executeSuspendFunctionAndRecordCancellationError` or `executeSuspendFunctionAndRecordCancellationException`

Reply:
I made a new commit with `executeSuspendFunctionAndRecordCancellationError`.